GUIDE.md

```diff
@@ -1,132 +1,281 @@
-import Control.Apply
-import Data.Functor
-import Data.Map
-import Routing
-import Routing.Match
-import Routing.Match.Class
-import Routing.Hash
-import Data.Int (floor)
+import Control.Alternative ((<|>))
+import Routing.Match (Match)
+import Routing.Match.Class (lit, int, str, end)
+```
+
+The available `Match` combinators are:
 
-data Locations
-  = Home
-  | Jupiter     -- The place where Boys go to get 'stupider'.
-  | Mars        -- Girls seek fame here.
-  | MPC Int     -- Minor Planet Circular (poor Pluto, et al.)
-  | Moon String -- They're not all made of cheese.
-  | PlanetWithAttributes (Map String String) -- Looking for a new home?
+* `lit` – Matches literal path segments. For example, `lit "posts"` would match
+  the path segment "posts" in our example URL.
+* `num` – Matches and returns a `Number` value.
+* `int` – Matches and returns an `Int` value.
+* `bool` – Matches and returns `true` or `false`.
+* `str` – Returns the path segment as is.
+* `param` – Extracts and returns a query parameter given a key.
+* `params` – Returns all query parameters.
+* `end` – Matches the end of the path.
 
+Lets define a route for `PostIndex`. This route has no parameters, so all we
+need to do is match the literal path segment "posts".
+
+```purescript
+postIndex :: Match Unit
+postIndex = lit "posts"
 ```
-We'll use `Locations` instead of `Routes` to avoid confusion with the `Route` (singular) type exposed by the library. But let's discuss the types later.  For now, we'll see it in action:
+
+However, this just yields a `Unit` value, and we need `MyRoute`. If there are
+no interesting values we want to consume, we can use `<$` from `Prelude`.
 
 ```purescript
-oneSlash :: Match Unit
-oneSlash = lit "/"
+postIndex :: Match MyRoute
+postIndex =
+  PostIndex <$ lit "posts
+```
 
-homeSlash :: Match Unit
-homeSlash = lit ""
+Our next routes require extracting an integer `PostId`.
 
-int :: Match Int
-int = floor <$> num
+```purescript
+post :: Match MyRoute
+post =
+  Post <$> lit "posts" *> int
 
-home :: Match Locations
-home = Home <$ oneSlash
+postEdit :: Match MyRoute
+postEdit =
+  PostEdit <$> lit "posts" *> int <* lit "edit"
+```
 
-jupiter :: Match Locations
-jupiter = Jupiter <$ (homeSlash *> lit "jupiter")
+Note the use of the `*>` and `<*` operators. These let us direct the focus of
+the value we want to consume. In `postEdit`, we want to consume the `int`,
+but we also need to match the "edit" suffix. The arrows point to the value we
+want.
 
-mars :: Match Locations
-mars = Mars <$ (homeSlash *> lit "mars" *> lit "hollywood")
+And now finally, we need to extract multiple segments for `PostBrowse`.
 
-mpc :: Match Locations
-mpc = MPC <$> (homeSlash *> lit "mpc" *> int)
+```purescript
+postBrowse :: Match MyRoute
+postBrowse =
+  PostBrowse <$> lit "posts" *> str <*> str
+```
 
-moon :: Match Locations
-moon = Moon <$> (homeSlash *> lit "moon" *> str)
+The `<*>` combinator has arrows on both sides because we want both values.
```

Comment:
> The <*> combinator has arrows on both sides because we want both values.

I like it!

Reply:
> Behold! Magically floating orb! 

Seriously, this lib uses finally-tagless encoding, applicative parsers, and effectful streams; do we think that anyone who doesn't know how `apply` work will use it?

Reply:
On the contrary, many people have come to PS wanting to write a simple web app, needing routing, all while having zero experience with Applicative parsers.

Reply:
And referring them to an applicative paper as background offers no help at all.

Reply:
Heh... I'm not suggesting this, all I'm just trying to say that this explanation: `two arrows <--> two values` - looks magical. Moreover there is one value and no arrows ¯\_(ツ)_/¯

But I think it's absolutely okay to leave it as is, because I can't come up with anything more descriptive

Reply:
Yes, clearly there is a bit of a leap, but this is just trying to give the user a nudge. Purescript-routing is clearly not the place to put a full-blown applicative parsing tutorial 😛